File: src/services/stakeholderService.js
/**
 * Stakeholder Identification & Analysis Service
 * Performs NLP-based extraction, categorization, and power/interest analysis
 */

class StakeholderService {
  constructor() {
    this.stakeholders = new Map()
    
    // Stakeholder type definitions
    this.STAKEHOLDER_TYPES = {
      'Product Owner': { power: 'High', interest: 'High', color: '#d4185d' },
      'Business Analyst': { power: 'High', interest: 'High', color: '#d4185d' },
      'Project Manager': { power: 'High', interest: 'High', color: '#d4185d' },
      'Developer': { power: 'Medium', interest: 'High', color: '#0066cc' },
      'QA Engineer': { power: 'Medium', interest: 'High', color: '#0066cc' },
      'Tester': { power: 'Medium', interest: 'High', color: '#0066cc' },
      'End User': { power: 'Low', interest: 'High', color: '#009900' },
      'Customer': { power: 'Medium', interest: 'High', color: '#0066cc' },
      'Stakeholder': { power: 'Medium', interest: 'Medium', color: '#ff9900' },
      'Manager': { power: 'High', interest: 'Medium', color: '#ff6600' },
      'Executive': { power: 'High', interest: 'Medium', color: '#ff6600' },
      'Sponsor': { power: 'High', interest: 'High', color: '#d4185d' }
    }

    // Extraction patterns
    this.EXTRACTION_PATTERNS = [
      /assigned to[:\s]+([A-Za-z\s]+?)(?:[,.]|$)/gi,
      /assignee[:\s]+([A-Za-z\s]+?)(?:[,.]|$)/gi,
      /stakeholder[s]?[:\s]+([A-Za-z\s,&]+?)(?:[,.]|$)/gi,
      /contact[:\s]+([A-Za-z\s]+?)(?:[,.]|$)/gi,
      /mentioned by[:\s]+([A-Za-z\s]+?)(?:[,.]|$)/gi,
      /owner[:\s]+([A-Za-z\s]+?)(?:[,.]|$)/gi,
      /lead[:\s]+([A-Za-z\s]+?)(?:[,.]|$)/gi,
      /manager[:\s]+([A-Za-z\s]+?)(?:[,.]|$)/gi,
      /reporter[:\s]+([A-Za-z\s]+?)(?:[,.]|$)/gi,
      /involving[:\s]+([A-Za-z\s,&]+?)(?:[,.]|$)/gi,
      /team[:\s]+([A-Za-z\s,&]+?)(?:[,.]|$)/gi,
      /department[:\s]+([A-Za-z\s,&]+?)(?:[,.]|$)/gi
    ]

    // Interest keywords
    this.INTEREST_KEYWORDS = {
      high: ['critical', 'important', 'urgent', 'required', 'must', 'essential', 'dependent', 'directly', 'directly involved'],
      medium: ['involved', 'contribute', 'support', 'helps', 'related', 'relevant'],
      low: ['inform', 'notify', 'aware', 'optional', 'nice to have', 'future']
    }

    // Power keywords
    this.POWER_KEYWORDS = {
      high: ['decides', 'approves', 'authorizes', 'budget', 'executive', 'sponsor', 'owner', 'lead', 'manager'],
      medium: ['implements', 'develops', 'designs', 'reviews', 'team lead', 'senior'],
      low: ['executes', 'supports', 'assists', 'junior', 'intern']
    }
  }

  /**
   * Identify stakeholders from tickets
   */
  identifyStakeholders(tickets) {
    const stakeholders = new Map()

    tickets.forEach((ticket, index) => {
      const text = `${ticket.summary || ''} ${ticket.description || ''} ${ticket.assignee || ''}`
      
      // Extract stakeholder mentions
      this.EXTRACTION_PATTERNS.forEach(pattern => {
        let match
        while ((match = pattern.exec(text)) !== null) {
          if (match[1]) {
            const names = match[1]
              .split(/[,&\s]+/)
              .map(n => n.trim())
              .filter(n => n.length > 2 && n.length < 50)

            names.forEach(name => {
              if (name && name.length > 2) {
                const normalized = this.normalizeName(name)
                
                if (!stakeholders.has(normalized)) {
                  stakeholders.set(normalized, {
                    name: this.formatName(name),
                    originalName: name,
                    mentions: [],
                    type: this.inferType(name),
                    power: 'Medium',
                    interest: 'Medium',
                    frequency: 0,
                    confidence: 0
                  })
                }

                const stakeholder = stakeholders.get(normalized)
                stakeholder.mentions.push({
                  ticketId: ticket.id,
                  context: text.substring(Math.max(0, match.index - 50), match.index + 100),
                  source: 'extraction'
                })
                stakeholder.frequency += 1
              }
            })
          }
        }
      })

      // Add assignee if present
      if (ticket.assignee && ticket.assignee !== 'Unassigned') {
        const normalized = this.normalizeName(ticket.assignee)
        
        if (!stakeholders.has(normalized)) {
          stakeholders.set(normalized, {
            name: this.formatName(ticket.assignee),
            originalName: ticket.assignee,
            mentions: [],
            type: 'Developer',
            power: 'Medium',
            interest: 'High',
            frequency: 0,
            confidence: 0.95
          })
        }

        const stakeholder = stakeholders.get(normalized)
        stakeholder.mentions.push({
          ticketId: ticket.id,
          context: 'Assigned to this ticket',
          source: 'assignee'
        })
        stakeholder.frequency += 1
      }
    })

    // Categorize each stakeholder
    stakeholders.forEach((stakeholder, key) => {
      this.categorizeStakeholder(stakeholder)
      // Calculate confidence based on mentions and patterns
      stakeholder.confidence = Math.min(0.95, 0.5 + (stakeholder.frequency * 0.1))
    })

    this.stakeholders = stakeholders
    return Array.from(stakeholders.values())
  }

  /**
   * Categorize stakeholder by power and interest levels
   */
  categorizeStakeholder(stakeholder) {
    let power = 'Medium'
    let interest = 'Medium'

    // Determine power level
    if (this.STAKEHOLDER_TYPES[stakeholder.type]) {
      power = this.STAKEHOLDER_TYPES[stakeholder.type].power
    } else {
      const name = (stakeholder.name || '').toLowerCase()
      const text = stakeholder.mentions.map(m => m.context || '').join(' ').toLowerCase()
      
      if (this.POWER_KEYWORDS.high.some(kw => name.includes(kw) || text.includes(kw))) {
        power = 'High'
      } else if (this.POWER_KEYWORDS.low.some(kw => name.includes(kw) || text.includes(kw))) {
        power = 'Low'
      }
    }

    // Determine interest level
    if (this.STAKEHOLDER_TYPES[stakeholder.type]) {
      interest = this.STAKEHOLDER_TYPES[stakeholder.type].interest
    } else {
      const text = stakeholder.mentions.map(m => m.context || '').join(' ').toLowerCase()
      
      if (this.INTEREST_KEYWORDS.high.some(kw => text.includes(kw))) {
        interest = 'High'
      } else if (this.INTEREST_KEYWORDS.low.some(kw => text.includes(kw))) {
        interest = 'Low'
      }
    }

    stakeholder.power = power
    stakeholder.interest = interest
    stakeholder.quadrant = this.getQuadrant(power, interest)
    stakeholder.color = this.getQuadrantColor(power, interest)

    return stakeholder
  }

  /**
   * Generate power/interest matrix data
   */
  generatePowerInterestMatrix(stakeholders = null) {
    const data = stakeholders || Array.from(this.stakeholders.values())

    const matrix = {
      highPowerHighInterest: [],
      highPowerLowInterest: [],
      lowPowerHighInterest: [],
      lowPowerLowInterest: [],
      summary: {
        total: data.length,
        byQuadrant: {},
        byType: {},
        averagePower: 0,
        averageInterest: 0
      }
    }

    // Categorize stakeholders
    data.forEach(stakeholder => {
      const quadrant = stakeholder.quadrant || this.getQuadrant(stakeholder.power, stakeholder.interest)
      
      switch (quadrant) {
        case 'manage':
          matrix.highPowerHighInterest.push(stakeholder)
          break
        case 'keep-satisfied':
          matrix.highPowerLowInterest.push(stakeholder)
          break
        case 'keep-informed':
          matrix.lowPowerHighInterest.push(stakeholder)
          break
        case 'monitor':
          matrix.lowPowerLowInterest.push(stakeholder)
          break
      }

      // Aggregate by type
      if (!matrix.summary.byType[stakeholder.type]) {
        matrix.summary.byType[stakeholder.type] = 0
      }
      matrix.summary.byType[stakeholder.type] += 1
    })

    // Calculate aggregates
    matrix.summary.byQuadrant = {
      manage: matrix.highPowerHighInterest.length,
      keep_satisfied: matrix.highPowerLowInterest.length,
      keep_informed: matrix.lowPowerHighInterest.length,
      monitor: matrix.lowPowerLowInterest.length
    }

    return matrix
  }

  /**
   * Analyze stakeholder influence network
   */
  analyzeInfluenceNetwork(stakeholders = null) {
    const data = stakeholders || Array.from(this.stakeholders.values())

    const nodes = data.map(s => ({
      id: this.normalizeName(s.name),
      label: s.name,
      type: s.type,
      power: s.power,
      interest: s.interest,
      frequency: s.frequency,
      color: s.color || this.getQuadrantColor(s.power, s.interest),
      size: Math.max(20, Math.min(50, 15 + s.frequency * 5))
    }))

    // Build relationships based on co-mentions
    const edges = []
    const mentioned = new Map()

    data.forEach(stakeholder => {
      stakeholder.mentions.forEach(mention => {
        if (!mentioned.has(mention.ticketId)) {
          mentioned.set(mention.ticketId, [])
        }
        mentioned.get(mention.ticketId).push(this.normalizeName(stakeholder.name))
      })
    })

    // Create edges between co-mentioned stakeholders
    mentioned.forEach((names, ticketId) => {
      for (let i = 0; i < names.length; i++) {
        for (let j = i + 1; j < names.length; j++) {
          const existingEdge = edges.find(e => 
            (e.from === names[i] && e.to === names[j]) || 
            (e.from === names[j] && e.to === names[i])
          )
          
          if (existingEdge) {
            existingEdge.weight += 1
          } else {
            edges.push({
              from: names[i],
              to: names[j],
              weight: 1
            })
          }
        }
      }
    })

    // Sort edges by weight
    edges.sort((a, b) => b.weight - a.weight)

    return {
      nodes,
      edges: edges.slice(0, 50), // Top 50 relationships
      network: {
        nodeCount: nodes.length,
        edgeCount: edges.length,
        density: edges.length / (nodes.length * (nodes.length - 1) / 2),
        topInfluencers: this.getTopInfluencers(data)
      }
    }
  }

  /**
   * Get top influencers by power and frequency
   */
  getTopInfluencers(stakeholders) {
    return stakeholders
      .map(s => ({
        ...s,
        influenceScore: (
          (s.power === 'High' ? 3 : s.power === 'Medium' ? 2 : 1) *
          (s.interest === 'High' ? 3 : s.interest === 'Medium' ? 2 : 1) *
          Math.log(s.frequency + 1)
        )
      }))
      .sort((a, b) => b.influenceScore - a.influenceScore)
      .slice(0, 10)
      .map(({ influenceScore, ...s }) => s)
  }

  /**
   * Get engagement recommendations
   */
  getEngagementRecommendations(matrix) {
    const recommendations = []

    // Manage closely (High Power, High Interest)
    if (matrix.highPowerHighInterest.length > 0) {
      recommendations.push({
        quadrant: 'manage',
        title: 'Manage Closely',
        stakeholders: matrix.highPowerHighInterest,
        strategies: [
          'Involve in key decisions and planning',
          'Regular updates and consultations',
          'Address concerns proactively',
          'Seek active participation'
        ],
        color: '#d4185d'
      })
    }

    // Keep Satisfied (High Power, Low Interest)
    if (matrix.highPowerLowInterest.length > 0) {
      recommendations.push({
        quadrant: 'keep-satisfied',
        title: 'Keep Satisfied',
        stakeholders: matrix.highPowerLowInterest,
        strategies: [
          'Regular updates with key information',
          'Highlight outcomes and benefits',
          'Avoid over-communication',
          'Ensure expectations are clear'
        ],
        color: '#ff6600'
      })
    }

    // Keep Informed (Low Power, High Interest)
    if (matrix.lowPowerHighInterest.length > 0) {
      recommendations.push({
        quadrant: 'keep-informed',
        title: 'Keep Informed',
        stakeholders: matrix.lowPowerHighInterest,
        strategies: [
          'Provide regular updates',
          'Include in relevant discussions',
          'Share progress and milestones',
          'Answer questions promptly'
        ],
        color: '#009900'
      })
    }

    // Monitor (Low Power, Low Interest)
    if (matrix.lowPowerLowInterest.length > 0) {
      recommendations.push({
        quadrant: 'monitor',
        title: 'Monitor',
        stakeholders: matrix.lowPowerLowInterest,
        strategies: [
          'General awareness only',
          'Minimal communication',
          'Watch for changes in interest/power',
          'Provide information if requested'
        ],
        color: '#cccccc'
      })
    }

    return recommendations
  }

  /**
   * Helper: Get quadrant name
   */
  getQuadrant(power, interest) {
    if (power === 'High' && interest === 'High') return 'manage'
    if (power === 'High' && interest === 'Low') return 'keep-satisfied'
    if (power === 'Low' && interest === 'High') return 'keep-informed'
    return 'monitor'
  }

  /**
   * Helper: Get quadrant color
   */
  getQuadrantColor(power, interest) {
    if (power === 'High' && interest === 'High') return '#d4185d' // Red
    if (power === 'High' && interest === 'Low') return '#ff6600' // Orange
    if (power === 'Low' && interest === 'High') return '#009900' // Green
    return '#cccccc' // Gray
  }

  /**
   * Helper: Infer stakeholder type from name
   */
  inferType(name) {
    const lower = name.toLowerCase()
    
    for (const [type, _] of Object.entries(this.STAKEHOLDER_TYPES)) {
      if (lower.includes(type.toLowerCase())) {
        return type
      }
    }
    
    return 'Stakeholder'
  }

  /**
   * Helper: Normalize name for comparison
   */
  normalizeName(name) {
    return name
      .toLowerCase()
      .trim()
      .replace(/\s+/g, '-')
  }

  /**
   * Helper: Format name for display
   */
  formatName(name) {
    return name
      .split(/[\s-]+/)
      .map(part => part.charAt(0).toUpperCase() + part.slice(1).toLowerCase())
      .join(' ')
  }

  /**
   * Validate stakeholder data
   */
  validateStakeholders(stakeholders) {
    const issues = []

    stakeholders.forEach((s, index) => {
      if (!s.name || s.name.trim().length === 0) {
        issues.push(`Stakeholder ${index}: Missing name`)
      }
      if (!['Low', 'Medium', 'High'].includes(s.power)) {
        issues.push(`Stakeholder ${s.name}: Invalid power level`)
      }
      if (!['Low', 'Medium', 'High'].includes(s.interest)) {
        issues.push(`Stakeholder ${s.name}: Invalid interest level`)
      }
      if (!s.mentions || s.mentions.length === 0) {
        issues.push(`Stakeholder ${s.name}: No mentions found (possible hallucination)`)
      }
    })

    return {
      valid: issues.length === 0,
      issues,
      warnings: this.detectPotentialHallucinations(stakeholders)
    }
  }

  /**
   * Detect hallucinated stakeholders
   */
  detectPotentialHallucinations(stakeholders) {
    const warnings = []

    stakeholders.forEach(s => {
      // Flag stakeholders with very low frequency
      if (s.frequency === 1 && s.confidence < 0.6) {
        warnings.push(`${s.name}: Single mention with low confidence (possible hallucination)`)
      }

      // Flag generic names
      if (s.name.length < 4 || s.name === 'User' || s.name === 'Team') {
        warnings.push(`${s.name}: Generic name (verify in source data)`)
      }

      // Flag malformed data
      if (s.name.includes(',') || s.name.includes('@') || s.name.length > 50) {
        warnings.push(`${s.name}: Suspicious format (verify extraction)`)
      }
    })

    return warnings
  }
}

export default new StakeholderService()

File: python-backend/services/stakeholder_service.py
"""Stakeholder identification and analysis service.

Implements stakeholder extraction, power/interest assessment, and matrix building.
"""
from __future__ import annotations

import re
from typing import Any, Dict, List, Optional, Set

from services.nlp_pipeline import StakeholderNLPPipeline

POWER_SCORES = {"Low": 1, "Medium": 2, "High": 3}


class StakeholderService:
    """Identifies and analyzes stakeholders from tickets."""

    # Power keywords for determining stakeholder influence
    POWER_KEYWORDS = {
        "executive": [
            "executive",
            "ceo",
            "cfo",
            "director",
            "president",
            "cto",
            "vp",
            "vice",
        ],
        "manager": [
            "manager",
            "lead",
            "supervisor",
            "coordinator",
            "owner",
            "stakeholder",
        ],
        "technical": [
            "developer",
            "engineer",
            "architect",
            "tech",
            "technical",
            "coding",
        ],
        "business": ["business", "analyst", "product", "owner", "sponsor"],
    }

    # Interest keywords
    INTEREST_KEYWORDS = {
        "high": [
            "critical",
            "must have",
            "required",
            "essential",
            "urgent",
            "blocker",
        ],
        "medium": [
            "important",
            "should have",
            "desired",
            "impacted",
            "involved",
        ],
        "low": [
            "optional",
            "nice to have",
            "could have",
            "future",
            "phase 2",
        ],
    }

    # Generic names to filter out
    GENERIC_NAMES = {
        "user",
        "admin",
        "manager",
        "developer",
        "team",
        "stakeholder",
        "client",
        "customer",
        "staff",
        "member",
        "person",
    }

    # Name extraction patterns
    EXTRACTION_PATTERNS = [
        re.compile(r"(?:assigned to|assignee|owner):\s*(\w+(?:\s+\w+)*)", re.IGNORECASE),
        re.compile(r"(?:reviewed by|approved by):\s*(\w+(?:\s+\w+)*)", re.IGNORECASE),
        re.compile(r"mentioned?\s+(\w+(?:\s+\w+)*)", re.IGNORECASE),
    ]

    def __init__(self, nlp_pipeline: Optional[StakeholderNLPPipeline] = None):
        """Initialize stakeholder service."""
        self.stakeholders: Dict[str, Dict[str, Any]] = {}
        self.extraction_patterns = self.EXTRACTION_PATTERNS
        self.nlp_pipeline = nlp_pipeline or StakeholderNLPPipeline()

    def identify_stakeholders(self, tickets: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Extract and analyze stakeholders from tickets.

        Args:
            tickets: List of ticket objects

        Returns:
            List of enriched stakeholder profiles
        """
        profiles: Dict[str, Dict[str, Any]] = {}

        for idx, ticket in enumerate(tickets):
            context = self._build_ticket_context(ticket)
            extracted_names = self._extract_names_from_text(context)

            # Process extracted names
            for name in extracted_names:
                normalized = self._normalize_name(name)
                profile = self._get_or_create_profile(profiles, normalized, name, context)

                profile["mentions"].append(
                    {
                        "ticketId": ticket.get("id"),
                        "index": idx,
                        "source": "extraction",
                        "context": context,
                        "snippet": self._get_snippet(context, name),
                    }
                )
                profile["frequency"] = len(profile["mentions"]) + len(profile["assignments"])

            # Process ticket assignee
            assignee = ticket.get("assignee")
            if assignee and assignee != "Unassigned":
                normalized = self._normalize_name(assignee)
                profile = self._get_or_create_profile(profiles, normalized, assignee, context)

                profile["assignments"].append(
                    {
                        "ticketId": ticket.get("id"),
                        "summary": ticket.get("summary", ""),
                        "status": ticket.get("status", "Unknown"),
                        "team": ticket.get("assigneeTeam"),
                        "role": ticket.get("assigneeRole", profile.get("type")),
                    }
                )
                profile["frequency"] = len(profile["mentions"]) + len(profile["assignments"])

        # Enrich profiles
        enriched_profiles = []
        for profile in profiles.values():
            combined_context = " ".join(m.get("context", "") for m in profile.get("mentions", []))

            profile["power"] = self._determine_power_level(profile, combined_context)
            profile["interest"] = self._determine_interest_level(profile, combined_context)
            profile["quadrant"] = self._get_quadrant(profile["power"], profile["interest"])
            profile["color"] = self._get_quadrant_color(profile["power"], profile["interest"])
            profile["influenceScore"] = self._calculate_influence_score(profile)
            profile["confidence"] = self._calculate_confidence(profile)
            profile["communicationPlan"] = self._build_communication_plan(profile)
            profile["originalNames"] = list(set(profile["originalNames"]))
            profile["roles"] = list(set(profile["roles"]))
            if profile["roles"]:
                profile["type"] = profile["roles"][0]

            # NLP enhancement (sentiment, entities, embeddings)
            profile = self._apply_nlp_enhancements(profile, combined_context)

            enriched_profiles.append(profile)

        self.stakeholders = {p["id"]: p for p in enriched_profiles}
        return enriched_profiles

    def _apply_nlp_enhancements(self, profile: Dict[str, Any], context: str) -> Dict[str, Any]:
        """Apply NLP enrichment to a stakeholder profile."""
        if not context or not self.nlp_pipeline:
            return profile

        try:
            profile = self.nlp_pipeline.enhance_profile(profile, context)
        except Exception as exc:  # pragma: no cover - optional dependency
            profile.setdefault("nlp", {})
            profile["nlp"]["error"] = str(exc)
        return profile

    def _build_ticket_context(self, ticket: Dict[str, Any]) -> str:
        """Combine ticket fields into single context string."""
        parts = [
            ticket.get("summary", ""),
            ticket.get("description", ""),
        ]

        if isinstance(ticket.get("acceptanceCriteria"), list):
            parts.append(" ".join(ticket["acceptanceCriteria"]))

        if isinstance(ticket.get("comments"), list):
            parts.append(" ".join(ticket["comments"]))

        assignee = ticket.get("assignee")
        if assignee:
            parts.append(f"Assignee: {assignee}")

        return " ".join(filter(None, parts))

    def _extract_names_from_text(self, text: str) -> List[str]:
        """Extract likely personal names from text using patterns.

        Args:
            text: Context text to search

        Returns:
            List of extracted names
        """
        candidates: Set[str] = set()

        if not text or not self.extraction_patterns:
            return list(candidates)

        for pattern in self.extraction_patterns:
            for match in pattern.findall(text):
                normalized = self._normalize_name(match)
                if normalized.lower() not in self.GENERIC_NAMES and len(normalized.split()) <= 3:
                    candidates.add(normalized)

        # Extract capitalized words adjacent to known stakeholder terms
        name_pattern = re.compile(r"([A-Z][a-z]+\s+[A-Z][a-z]+)")
        for match in name_pattern.findall(text):
            normalized = self._normalize_name(match)
            if normalized.lower() not in self.GENERIC_NAMES:
                candidates.add(normalized)

        return list(candidates)

    def _split_candidate_names(self, raw: str) -> List[str]:
        """Split multi-name strings by common delimiters."""
        names = re.split(r"(?:,|&|\band\b|\+|/)+", raw, flags=re.IGNORECASE)
        return [n.strip() for n in names if n.strip()]

    def _is_likely_person_name(self, name: str) -> bool:
        """Check if string is likely a personal name."""
        trimmed = name.strip()

        if not trimmed or len(trimmed) < 2:
            return False

        if len(trimmed) > 100:
            return False

        if trimmed.lower() in self.GENERIC_NAMES:
            return False

        # Filter numeric-only strings
        if trimmed.isdigit():
            return False

        return True

    def _normalize_name(self, name: str) -> str:
        """Normalize name for matching and aggregation."""
        return name.strip().lower().replace("_", " ")

    def _get_or_create_profile(
        self, profiles: Dict[str, Dict], norm_id: str, original_name: str, context: str
    ) -> Dict[str, Any]:
        """Get existing profile or create new one."""
        if norm_id not in profiles:
            profiles[norm_id] = {
                "id": norm_id,
                "name": original_name,
                "originalNames": [original_name],
                "mentions": [],
                "assignments": [],
                "frequency": 0,
                "type": self._infer_role(original_name, context),
                "roles": [self._infer_role(original_name, context)],
                "power": "Medium",
                "interest": "Medium",
                "quadrant": "middle",
                "confidence": 0.5,
                "influenceScore": 0.5,
            }
        else:
            profiles[norm_id]["originalNames"].append(original_name)

        return profiles[norm_id]

    def _infer_role(self, name: str, context: str) -> str:
        """Infer stakeholder role from name and context."""
        combined = f"{name} {context}".lower()

        for role, keywords in self.POWER_KEYWORDS.items():
            if any(kw in combined for kw in keywords):
                return role

        return "stakeholder"

    def _determine_power_level(self, profile: Dict[str, Any], context: str) -> str:
        """Determine stakeholder power level."""
        combined = f"{profile['name']} {context}".lower()

        exec_score = sum(1 for kw in self.POWER_KEYWORDS.get("executive", []) if kw in combined)
        manager_score = sum(1 for kw in self.POWER_KEYWORDS.get("manager", []) if kw in combined)
        tech_score = sum(1 for kw in self.POWER_KEYWORDS.get("technical", []) if kw in combined)

        if exec_score >= 2:
            return "High"
        elif manager_score >= 2 or (exec_score >= 1 and manager_score >= 1):
            return "High"
        elif tech_score >= 3:
            return "Medium"
        elif profile.get("frequency", 0) > 5:
            return "Medium"

        return "Low"

    def _determine_interest_level(self, profile: Dict[str, Any], context: str) -> str:
        """Determine stakeholder interest level."""
        combined = f"{profile['name']} {context}".lower()

        high_score = sum(1 for kw in self.INTEREST_KEYWORDS.get("high", []) if kw in combined)
        medium_score = sum(1 for kw in self.INTEREST_KEYWORDS.get("medium", []) if kw in combined)

        if high_score >= 2:
            return "High"
        elif high_score >= 1 and medium_score >= 1:
            return "High"
        elif medium_score >= 2 or profile.get("frequency", 0) > 3:
            return "Medium"

        return "Low"

    def _get_quadrant(self, power: str, interest: str) -> str:
        """Determine quadrant based on power and interest."""
        power_idx = POWER_SCORES.get(power, 2)
        interest_idx = POWER_SCORES.get(interest, 2)

        if power_idx >= 2 and interest_idx >= 2:
            return "highPowerHighInterest"
        elif power_idx >= 2 and interest_idx < 2:
            return "highPowerLowInterest"
        elif power_idx < 2 and interest_idx >= 2:
            return "lowPowerHighInterest"
        else:
            return "lowPowerLowInterest"

    def _get_quadrant_color(self, power: str, interest: str) -> str:
        """Get color for quadrant visualization."""
        quadrant = self._get_quadrant(power, interest)
        colors = {
            "highPowerHighInterest": "#1e40af",  # Blue
            "highPowerLowInterest": "#7c3aed",  # Purple
            "lowPowerHighInterest": "#059669",  # Green
            "lowPowerLowInterest": "#6b7280",  # Gray
        }
        return colors.get(quadrant, "#9ca3af")

    def _calculate_influence_score(self, profile: Dict[str, Any]) -> float:
        """Calculate overall influence score."""
        power_score = POWER_SCORES.get(profile.get("power", "Low"), 1) / 3.0
        frequency_score = min(profile.get("frequency", 0) / 10.0, 1.0)
        assignment_score = min(len(profile.get("assignments", [])) / 5.0, 1.0)

        return (power_score * 0.5 + frequency_score * 0.25 + assignment_score * 0.25)

    def _calculate_confidence(self, profile: Dict[str, Any]) -> float:
        """Calculate extraction confidence."""
        frequency = profile.get("frequency", 0)
        if frequency >= 5:
            return 0.95
        elif frequency >= 3:
            return 0.80
        elif frequency >= 2:
            return 0.65
        return 0.5

    def _build_communication_plan(self, profile: Dict[str, Any]) -> Dict[str, Any]:
        """Build communication strategy for stakeholder."""
        quadrant = profile.get("quadrant")
        strategies = {
            "highPowerHighInterest": {
                "strategy": "Manage Closely",
                "frequency": "Weekly",
                "channels": ["Meeting", "Email", "Presentation"],
                "involvement": "High",
            },
            "highPowerLowInterest": {
                "strategy": "Keep Satisfied",
                "frequency": "Bi-weekly",
                "channels": ["Email", "Report"],
                "involvement": "Medium",
            },
            "lowPowerHighInterest": {
                "strategy": "Keep Informed",
                "frequency": "Weekly",
                "channels": ["Email", "Update", "Chat"],
                "involvement": "Medium",
            },
            "lowPowerLowInterest": {
                "strategy": "Monitor",
                "frequency": "Monthly",
                "channels": ["Email"],
                "involvement": "Low",
            },
        }
        return strategies.get(quadrant, strategies["lowPowerLowInterest"])

    def _get_snippet(self, context: str, name: str) -> str:
        """Extract snippet around name mention."""
        lower_context = context.lower()
        lower_name = name.lower()
        idx = lower_context.find(lower_name)

        if idx == -1:
            return context[:100]

        start = max(0, idx - 30)
        end = min(len(context), idx + len(name) + 30)
        return context[start:end].strip()
